### ISSUE TITLE
Implement `weakhash*` primitives

### ISSUE THREAD

**Issue by vishesh** [vishesh/racketscript#106]
As discussed in #105

**Comment by stchang** [comment]
Running into some problems trying to implement `make-weak-hash` (the one that uses `equal?`).

Specifically, I can't figure out how to do a `hash-ref` when the given key is `equal?` but not `eq?` to one that is (or was) in the hash.

In this case, `hash-ref` should return the corresponding value if the key has not been collected yet but how can I figure out if the key has been collected?

Holding the keys in a `WeakMap` or `WeakSet` doesnt work since those are not iterable, and I cant use `has` since that compares with `eq?`.

(The fact that `WeakMaps` are not iterable also makes it hard to print a `weak-hash`, but that's a separate and more minor issue.)

Am I missing something obvious? Any suggestions?

**Comment by vishesh** [comment]
I don't think you are missing anything. I'm convinced myself it is not possible with following reasoning.

`WeapMap` works on objects, and essentially what we want is to compute a unique identify for all values that are `equal?` at that point of execution. Hashing has two problems:

1. It generates a number for which we have to use a regular `Map`.
2. Even if `WeakMap` can take integers, we can't guarantee a unique hash for `equal?` values, regardless how many hashes you take (only make collisions less frequent). And like you said, to check for collisions we need iteration through another WeakMap collection mapped to the key, which we again can't do.

So, we need a mechanism to map our all `equal?` values to a unique identity object, which can be used when we do ref. Building such map is again equivalent to the original problem we have at hand. Overall, if a value is in WeakMap, one object of same value should be in a regular value based hash/set with which we have leak. We won't leak everything, i.e. out of 10 objects involved that are equal 1 would stay in memory forever, but it is still not correct.

**Comment by stchang** [comment]
Ok I will implement `make-weak-hasheq` with `WeakMap`.

For `make-weak-hasheqv`, I think it could be a `WeakMap` but it depends on how we implement `eqv`? Specifically, are there number objects that are `eqv?` but not `eq?`.

For `make-weak-hash` is it better to error or default to non-weak hash (and leak memory)?

**Comment by vishesh** [comment]
> For make-weak-hash is it better to error or default to non-weak hash (and leak memory)?

Ideally, I think it should error to discourage any use of it in RacketScript progra m. However, if the Racket standard library is using somewhere that is commonly used, it can be a problem. Right now, the most useful occurrence I see is in `string-replace` in `racket/string :((. Maybe do a`console.warn` when doing this? In the end, `string-replace` using a weakhash in long-running application is impractical (On side note, we need to think about cases like these in standard library in general).

> are there number objects that are eqv? but not eq?.

Numbers are currently `eq?`. There are two problems though:

1. Numbers can't be keys in `WeakMap`. So current numbers can't be in our WeakMap.
2. If we implement Racket style numbers, they no longer will be `eq?`. But should be fine as `eq?` can be implementation specific?

> For make-weak-hasheqv, I think it could be a WeakMap but it depends on how we implement eqv? 

Yeah right. We can make `eqv?` work as Racket does, but I don't think we can actually make `make-weak-hasheqv` because of the same reason as for the `make-weak-hash`. We will have multiple values with different object but same equality.

**Comment by stchang** [comment]
I like the warning suggestion. I'll go with that.

For weak-hasheq and weak-hasheqv I'm implementing it using @glebm's suggestion and keeping both a `WeakMap` for objects and regular `Map` for non-objects. So all the operations will need a `if typeof (key) === object` but I think that should work.

**Comment by vishesh** [comment]
Cool!

**Comment by glebm** [comment]
Ouch, I hadn't realized WeakMap/Set are not iterable. I agree that this means `eqv` and `equal` versions are basically impossible.

There is a stage 1 [tc39 proposal for weak references](https://github.com/tc39/proposal-weakrefs/) but it doesn't look like it will reach the next stage anytime soon.

**Comment by glebm** [comment]
The weak references TC39 proposal is now stage 2: https://github.com/tc39/proposal-weakrefs, meaning "The committee expects the feature to be developed and eventually included in the standard".

**Comment by stchang** [comment]
I think `weak-hash` and `weak-box`es should be doable now with [WeakRefs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef)

Not sure about ephemerons though.

**Comment by vishesh** [comment]
There is also WeakMap and WeakSet, so maybe those can be used to implemented weak-hash instead without ephemerons.

**Comment by stchang** [comment]
Yes, `WeakHash` didn't work because we couldn't iterate. But manually using WeakRefs should work

**Comment by vishesh** [comment]
I don't think anyone is working on it. Go ahead!

**Comment by gamburgm** [comment]
Wanted to suggest I what I was thinking here before I dive in too deep:

It seems like this should be doable, but will require re-implementing portions of the `HAMT` implementation in Javascript (or re-implementing a few functions for weak hashes). 

Reason being: we cannot modify the Javascript garbage collector to delete an entry in the hash when the key is garbage collected, instead inserting a level of indirection using a `WeakRef`. As a consequence, we'll still be leaking hash entries, because we won't remove an entry when the key is garbage collected.

Instead, we have to iterate through the hash and remove entries that have had GC'ed keys ourselves whenever the user invokes a function that could be affected by leaked keys/entries.

For example, imagine someone inserts some key-value pair into a weak hash, then gets the count of the hash. Unbeknownst to the developer, Javascript garbage collects between those two function calls, and the key is deleted (but not the `WeakRef` to it). The count will consequently be greater than the count of elements whose keys haven't been garbage collected.

To return the as-close-as-possible result, we'd want to iterate through the keys and remove any entries whose keys have been collected, then return the count. The same applies for other similarly-affected functions, e.g. `fold` would require that we remove elements as we iterate through the hash. 

We likely can't eliminate inconsistencies entirely, but we can shrink it to virtually non-existent.

**Comment by stchang** [comment]
> It seems like this should be doable, but will require re-implementing portions of the `HAMT` implementation in Javascript (or re-implementing a few functions for weak hashes).
> 
> Reason being: we cannot modify the Javascript garbage collector to delete an entry in the hash when the key is garbage collected, instead inserting a level of indirection using a `WeakRef`. As a consequence, we'll still be leaking hash entries, because we won't remove an entry when the key is garbage collected.
> 
> Instead, we have to iterate through the hash and remove entries that have had GC'ed keys ourselves whenever the user invokes a function that could be affected by leaked keys/entries.

That's a good point, thanks.

I think that we should not be doing manual memory management tasks like that. Firstly, because it will be difficult to test that our implementation is even correct, but also because it opens the door to many more corner cases that we're not thinking of.

I think that now that we've explored many options, at this point we should just compile `weak-hash` to `WeakMaps`.

The only thing we'll lose (I think) is the iteration functions in the api, but programmers shouldnt be using those with weak maps anyways because they're unpredictable (which is why js doesnt allow iteration on WeakMaps in the first place).

@vishesh  @samth Any opinion on this?

**Comment by samth** [comment]
I think there's significant racket code that uses iteration and other primitives on weak maps, probably including the expander.

**Comment by stchang** [comment]
After discussion, I'm dropping the "small project tag"

**Comment by gamburgm** [comment]
With `FinalizationRegistry` now supported, iterable `weak-hash` and `ephemeron-hash` should both be relatively straightforward to implement. An `ephemeron-hash` is effectively implemented [here](https://github.com/tc39/proposal-weakrefs#iterable-weakmaps). 

With that in mind, I'll probably take a stab at it sometime later this week.

**Comment by stchang** [comment]
For reference: 
https://github.com/ocsigen/js_of_ocaml/issues/1266
https://github.com/ocsigen/js_of_ocaml/issues/1038
